### PR TITLE
Issue #2793205 by mglaman: Price element has #maxlength, causes error…

### PR DESCRIPTION
**Pull request by mglaman**
… in Unicode::strlen

### CODE REVIEW

**Comment by mglaman** [comment]
@bojanz when #maxlength is set it gets passed to Unicode::strlen, but the value of the element is an array. so ['amount' => '9.99', 'currency_code' => 'USD'] when can't be `strlen`'d

See https://github.com/drupalcommerce/commerce/pull/478 which exposes this fact.

**Comment by mglaman** [comment]
Lets just put it here instead of on element definition

`'#maxlength' => $element['#maxlength'],`

File: modules/price/src/Element/Price.php [hunk 1]
```diff
@@ -27,16 +27,16 @@
  *
  * @FormElement("commerce_price")
  */
-class Price extends Textfield {
+class Price extends FormElement {
```
Comment (mglaman) [inline modules/price/src/Element/Price.php:30]: @bojanz I looked at `\Drupal\Core\Render\Element\Textfield` and we weren't gaining anything out of extending it.  We weren't using any of the parent class's process or pre-render methods.

File: modules/price/src/Element/Price.php [hunk 2]
```diff
@@ -128,8 +128,8 @@ public static function processElement(array $element, FormStateInterface $form_s
       '#title' => $element['#title'],
       '#default_value' => $amount,
       '#required' => $element['#required'],
-      '#size' => $element['#size'],
-      '#maxlength' => $element['#maxlength'],
+      '#size' => $element['#amount_size'],
+      '#maxlength' => $element['#amount_maxlength'],
```
Comment (mglaman) [inline modules/price/src/Element/Price.php]: Allows us to control size, maxlength in main element and not hardcode. But also not trigger the `Unicode::strlen()`

File: modules/price/src/Element/Price.php (hunk 1 above)
Comment (bojanz) [reply modules/price/src/Element/Price.php:30]: Sure.

File: modules/price/src/Element/Price.php [hunk 3]
```diff
@@ -27,16 +27,16 @@
...
 
   /**
    * {@inheritdoc}
    */
   public function getInfo() {
     $class = get_class($this);
     return [
-      '#size' => 10,
-      '#maxlength' => 128,
+      '#amount_size' => 10,
```
Comment (bojanz) [inline modules/price/src/Element/Price.php]: Can't we not still use #size and #maxlength and unset it after transfering it to the subelement? I believe that's what other complex form elements do?

Comment (mglaman) [reply modules/price/src/Element/Price.php]: I didn't see any other patterns like that, here's the check in question `\Drupal\Core\Form\FormValidator::performRequiredValidation`.

I'll just unset after we pass to amount

File: modules/price/src/Element/Price.php [hunk 4]
```diff
@@ -134,6 +134,9 @@ public static function processElement(array $element, FormStateInterface $form_s
       // separator to use. This is the same pattern Drupal core uses.
       '#placeholder' => $number_formatter->format('9.99'),
     ];
+    unset($element['#size']);
```
Comment (mglaman) [inline modules/price/src/Element/Price.php:137]: @bojanz done

**Pull request merged**